### PR TITLE
Submission: Recursive Layer Sharing | 13.9 MB | 1.53 BPB

### PR DESCRIPTION
**Pull request by negrurv**
Hello! Here is my submission for the Parameter Golf challenge.

Architecture Hack:
I completely removed the standard encoder/decoder split and implemented a "Recursive Transformer." The model initializes a single Block of weights and recursively passes the data through that exact same memory block 9 times during the forward pass. I also implemented a weightless RMSNorm (RMSNormNoWeight) to shave off extra kilobytes.

Final Stats (Local H100 Testing):

Model Dimension: 512

Total Compressed Size: 13.9 MB

Validation BPB: 1.5363

Train Time: Completed full 600s wallclock

Thanks for hosting the challenge!

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Submission: Recursive Layer Sharing | 13.9 MB | 1.53 BPB

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

PR #1542 modifies `train_gpt.py` with two parameter-reduction techniques, both purely architectural:

**1. Weight-sharing across transformer layers (lines 684–694)**
A single `Block` instance (`self.shared_block`) is created and referenced `num_layers` (9) times via `nn.ModuleList`. All 9 forward passes reuse the same weights. This is a standard recursive/tied-weight technique — no lookup tables, no hash keys, no target leakage.

```python
self.shared_block = Block(model_dim, num_heads, num_kv_heads, mlp_mult, rope_base, qk_gain_init)
self.blocks = nn.ModuleList([self.shared_block for _ in range(num_layers)])
```

**2. RMSNormNoWeight for the final norm (lines 647–659, 697)**
Replaces the final `RMSNorm` (which has a learnable per-dim scale) with `RMSNormNoWeight` — a plain RMS normalization with no parameters. Saves `model_dim` parameters.

**Illegal pattern checks — all clear:**

- **N-gram family bug**: No hash tables, no `ctx_hash`, no `primes[k]`, no XOR of target into a lookup key anywhere in the file.
- **Pre-Quant TTT / multi-epoch val_tokens AdamW**: No test-time training, no optimizer steps on validation tokens, no epoch loops over val data.
- **Scored-region SLOT**: No mask-optimize-score pattern on the same region.

**Training loop (lines 985–1067)**: Standard — train data only for gradient steps, `eval_val` called under `torch.inference_mode()` (line 250) for validation, no optimizer interaction with val tokens at any point.

The submission is a clean pure-neural entry: it reduces parameter count via weight tying across layers and dropping the final norm's learnable scale, with no algorithmic shortcuts that touch validation data during training.

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
